### PR TITLE
Implement Backups

### LINKED ISSUE #418: Implement AWS Backup resources
AWS Backup enables you to centralize and automate data protection across AWS services.

The following is a rough mapping of the AWS Backup SDK to Gyro resources.

**Backup Selection**

```
aws::backup-selection production
    name: production
    plan-id: my-plan-id
    iam-role: $(aws::iam-role my-backup-role)

    resources: [$rds-arn, $efs-arn]
    tags: {
        backup-policy: production
    }
end
```

**Backup Vault**

```
aws::backup-vault my-vault
    name: my-vault
    encryption-key-arn: $arn
    tags: {
        tag1: value1
    }
end
```

**Backup Plan**

```
aws::backup-plan backup
    name: my-rds-backups

    rule
        name: weekly
        target-backup-vault: $(aws::backup-vault my-vault)

        schedule: "cron(0 5 ? * * *)"
        start-window-minutes: 10

        completion-window-minutes: 60
        enable-continous-backup: false

        recovery-point-tags: {
           tag1: value1
        }

        lifecycle
            delete-after-days: 365
            move-to-coldstorage-after-days: 30
        end

        copy-action
            destination-backup-value-arn: arn
            lifecycle
                delete-after-days: 365
                move-to-coldstorage-after-days: 30
            end
        end
    end

    tags: {
        project: "my project"
    }
end
```

### PR DESCRIPTION
**Pull request by harjain99**
Fixes #418

### CODE REVIEW

File: examples/backup/backup.gyro [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+    configuration
+        name: production
+        role: $(external-query aws::iam-role { name: 'AWSBackupDefaultServiceRole'})
+        resources: ["arn:aws:rds:us-east-2:242040583208:db:aurora-mysql-cluster"]
+    end
+end
+
+aws::kms-key kms-example
+    aliases: ["alias/backup-example"]
```
Comment (deepanjan90) [inline examples/backup/backup.gyro:27]:
```suggestion
    aliases: ["alias/backup-selection-example"]
```

File: examples/backup/backup.gyro [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    bypass-policy-lockout-safety-check: "true"
+    description: "sample kms key"
+    enabled: "true"
+    key-rotation: "true"
+    policy: "kms-policy.json"
+    tags: {
+        Name: "kms-example"
```
Comment (deepanjan90) [inline examples/backup/backup.gyro:34]:
```suggestion
        Name: "kms-for-backup-selection-example"
```

**Review by deepanjan90 — CHANGES_REQUESTED** [review]
1st Pass
 - Inline comments
 - Add License snippet to all java files
 - Use paginators for the finders
 - Add packgeinfo file to the package

File: src/main/java/gyro/aws/backup/BackupCopyAction.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        if (model.lifecycle() != null) {
+            BackupLifecycle backupLifecycle = newSubresource(BackupLifecycle.class);
+            backupLifecycle.copyFrom(model.lifecycle());
+            setLifecycle(backupLifecycle);
+        }
+    }
+
+    @Override
+    public String primaryKey() {
+        return String.format("Vault: %s, Lifecycle: (%s,  %s)", getDestinationBackupVault().getName(),
```
Comment (deepanjan90) [inline src/main/java/gyro/aws/backup/BackupCopyAction.java:67]:
```suggestion
        return String.format("with vault: %s and lifecycle: (%s,  %s)", getDestinationBackupVault().getName(),
```

File: src/main/java/gyro/aws/backup/BackupCondition.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    public void copyFrom(Condition model) {
+        setConditionKey(model.conditionKey());
+        setConditionType(model.conditionType());
+        setConditionValue(model.conditionValue());
+    }
+
+    @Override
+    public String primaryKey() {
+        return String.format(
+            "Type: %s, Key: %s, Value: %s",
```
Comment (deepanjan90) [inline src/main/java/gyro/aws/backup/BackupCondition.java:77]:
```suggestion
            "with type: %s, key: %s and value: %s",
```

File: src/main/java/gyro/aws/backup/BackupCondition.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+import gyro.aws.Copyable;
+import gyro.core.resource.Diffable;
+import gyro.core.validation.Required;
+import software.amazon.awssdk.services.backup.model.Condition;
+import software.amazon.awssdk.services.backup.model.ConditionType;
+
+public class BackupCondition extends Diffable implements Copyable<Condition> {
+
+    private String conditionKey;
```
Comment (deepanjan90) [inline src/main/java/gyro/aws/backup/BackupCondition.java:27]:
```suggestion
    private String key;
```

File: src/main/java/gyro/aws/backup/BackupCondition.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    private ConditionType conditionType;
```
Comment (deepanjan90) [inline src/main/java/gyro/aws/backup/BackupCondition.java:28]:
```suggestion
    private ConditionType type;
```

File: src/main/java/gyro/aws/backup/BackupCondition.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    private String conditionValue;
```
Comment (deepanjan90) [inline src/main/java/gyro/aws/backup/BackupCondition.java:29]:
```suggestion
    private String value;
```

File: src/main/java/gyro/aws/backup/AdvancedBackupSetting.java [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+    @Override
+    public void copyFrom(software.amazon.awssdk.services.backup.model.AdvancedBackupSetting model) {
+        setBackupOptions(model.backupOptions());
+        setResourceType(model.resourceType());
+    }
+
+    @Override
+    public String primaryKey() {
+        return String.format("Resource Type: %s, Options: (%s)", getResourceType(), getBackupOptions().toString());
```
Comment (deepanjan90) [inline src/main/java/gyro/aws/backup/AdvancedBackupSetting.java:73]:
```suggestion
        return String.format("with resource type: %s and options: (%s)", getResourceType(), getBackupOptions().toString());
```

File: src/main/java/gyro/aws/backup/BackupRule.java [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+                BackupCopyAction backupCopyAction = newSubresource(BackupCopyAction.class);
+                backupCopyAction.copyFrom(c);
+                return backupCopyAction;
+            }).collect(Collectors.toList()));
+        }
+    }
+
+    @Override
+    public String primaryKey() {
+        return String.format("Rule name: %s", getName());
```
Comment (deepanjan90) [inline src/main/java/gyro/aws/backup/BackupRule.java:193]:
```suggestion
        return getName();
```

File: src/main/java/gyro/aws/backup/BackupLifecycle.java [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+import gyro.aws.Copyable;
+import gyro.core.resource.Diffable;
+import gyro.core.resource.Updatable;
+import software.amazon.awssdk.services.backup.model.Lifecycle;
+
+public class BackupLifecycle extends Diffable implements Copyable<Lifecycle> {
+
+    private Long deleteAfterDays;
+    private Long moveToColdStorageAfterDays;
```
Comment (deepanjan90) [inline src/main/java/gyro/aws/backup/BackupLifecycle.java:27]: What if a user does not give any value ?

**Review by deepanjan90 — CHANGES_REQUESTED** [review]
2nd pass
 -Inline changes
 - Add @subresource doc tags

File: src/main/java/gyro/aws/backup/BackupLifecycle.java (hunk 10 above)
Comment (harjain99) [reply src/main/java/gyro/aws/backup/BackupLifecycle.java:27]: It never transitions to cold storage.

**Review by deepanjan90 — APPROVED** [review]

**Pull request merged**